modules/renter/hostdb/hostweight.go

```diff
@@ -41,104 +128,227 @@ func (hdb *HostDB) calculateHostWeight(entry modules.HostDBEntry) types.Currency
+// function.
+func (hdb *HostDB) uptimeAdjustments(entry modules.HostDBEntry) float64 {
+	// Special case: if we have scanned the host twice or fewer, don't perform
+	// uptime math.
+	if len(entry.ScanHistory) == 0 {
+		return 0.25
+	}
+	if len(entry.ScanHistory) == 1 {
+		if entry.ScanHistory[0].Success {
+			return 0.75
+		}
+		return 0.25
+	}
+	if len(entry.ScanHistory) == 2 {
+		if entry.ScanHistory[0].Success && entry.ScanHistory[1].Success {
+			return 0.85
+		}
+		if entry.ScanHistory[0].Success || entry.ScanHistory[1].Success {
+			return 0.50
+		}
+		return 0.05
+	}
+
+	// Compute the total measured uptime and total measured downtime for this
+	// host.
+	var uptime time.Duration
+	var downtime time.Duration
+	recentTime := entry.ScanHistory[0].Timestamp
+	recentSuccess := entry.ScanHistory[0].Success
+	for _, scan := range entry.ScanHistory[1:] {
+		if recentTime.After(scan.Timestamp) {
+			hdb.log.Critical("Host entry scan history not sorted.")
+		}
+		if recentSuccess {
+			uptime += scan.Timestamp.Sub(recentTime)
+		} else {
+			downtime += scan.Timestamp.Sub(recentTime)
+		}
+		recentTime = scan.Timestamp
+		recentSuccess = scan.Success
+	}
+	// Sanity check against 0 total time.
+	if uptime == 0 && downtime == 0 {
+		return 0.001 // Shouldn't happen.
+	}
+
+	// Compute the uptime ratio, but shift by 0.02 to acknowledge fully that
+	// 98% uptime and 100% uptime is valued the same.
+	uptimeRatio := float64(uptime) / float64(uptime+downtime)
+	if uptimeRatio > 0.98 {
+		uptimeRatio = 0.98
+	}
+	uptimeRatio += 0.02
+
+	// Cap the total amount of downtime allowed based on the total number of
+	// scans that have happened.
+	allowedDowntime := 0.03 * float64(len(entry.ScanHistory))
+	if uptimeRatio < 1-allowedDowntime {
+		uptimeRatio = 1 - allowedDowntime
+	}
+
+	// Calculate the penalty for low uptime. Penalties increase extremely
+	// quickly as uptime falls away from 95%.
 	//
-	// The renter has control over how much collateral the host uses.
-	// Currently, this control is not implemented, so the hosts are penalized
-	// for setting very high collateral values. Once the renter is clamping the
-	// amount being spent on collateral, the hostdb can also clamp the amount
-	// of collateral being taken into account by the host, to optimize the
-	// host's score for the renter's needs.
-	if entry.Collateral.IsZero() {
-		// Instead of zeroing out the weight, just return the weight as though
-		// the collateral is 1 hasting. Competitively speaking, this is
-		// effectively zero.
-		return weight
-	}
-	weight = weight.Mul(entry.Collateral)
+	// 100% uptime = 1
+	// 98%  uptime = 1
+	// 95%  uptime = 0.91
+	// 90%  uptime = 0.51
+	// 85%  uptime = 0.16
+	// 80%  uptime = 0.03
+	// 75%  uptime = 0.005
+	// 70%  uptime = 0.001
+	// 50%  uptime = 0.000002
+	uptimePenalty := float64(1)
+	for i := float64(1); i > uptimeRatio && i > 0.80; i -= 0.01 {
+		uptimePenalty *= uptimeRatio
+	}
```

Comment:
originally this was happening with types.Currency instead of float. I guess I also didn't realize the math package operated on floats.